synapse/federation/sender/per_destination_queue.py

```diff
@@ -542,22 +550,53 @@ async def _catch_up_transmission_loop(self) -> None:
                     self._destination, last_successful_stream_ordering
                 )
 
-    def _get_rr_edus(self, force_flush: bool) -> Iterable[Edu]:
+    def _get_rr_edus(self, force_flush: bool, limit: int) -> Iterable[Edu]:
         if not self._pending_rrs:
             return
         if not force_flush and not self._rrs_pending_flush:
             # not yet time for this lot
             return
 
-        edu = Edu(
-            origin=self._server_name,
-            destination=self._destination,
-            edu_type=EduTypes.RECEIPT,
-            content=self._pending_rrs,
-        )
+        # Build the EDUs needed to send these receipts. This is a bit complicated
+        # since we can share one for each unique (room, receipt type, user), but
+        # need additional ones for different threads. The result is that we will
+        # send N EDUs where N is the number of unique threads across rooms.
+        #
+        # This could be more efficient by bundling users who have sent receipts
+        # for different threads.
```

Comment:
I wonder if we need to care about this more now that we're enforcing a limit?

Reply:
That's a good point. eg. if we have receipts (thread1, user1), (thread2, user2), (thread3, user3) all for the same (room_id, receipt_type), we'd send out 3 EDUs right now. Whereas it's possible to bundle them all into a single EDU. I think we do need to care about this unfortunately.

It's also worth noting that we choose which thread's receipts to send using `popitem()` right now, which always returns the most recent thread that was added, and not the one that's been sitting in the queue the longest. So if there's a very busy thread in a room, receipts for other threads might not make it out (even though they could, in theory, if we bundled receipts more densely).

Reply:
OK. I'll need to think a bit more about how to arrange the data to make this possible then!

Reply:
I reworked this to build the EDUs when we add receipts instead of here.

I think this works better, although there is a downside still that we will return the oldest added thread, even if it is has been updated more recently. I don't see this as necessarily bad though. E.g.:

1. Thread A is read
2. Thread B is read in the same room (a second EDU gets added)
3. Thread A is read again at a later event.
4. EDUs are sent, so Thread A is sent first.